### PR TITLE
Fix MaxMeasuredValue is out of range Temperature Measurement Cluster

### LINKED ISSUE #9864: Temperature Measurement Cluster - mandatory attribute: MaxMeasuredValue is out of range
Spec Snapshot : 
Screen Shot 2021-09-21 at 10 30 36 AM

MaxMeasuredValue needs to be between -27314 to 32767 , but SDK is reporting -32768.

[1632245964135] [87111:7198158] CHIP: [TOO] Temperature Measurement - read the mandatory attribute: MaxMeasuredValue: Success Response
[1632245964135] [87111:7198158] CHIP: [TOO] Warning: maxMeasuredValue type checking is not implemented yet. Expected type: 'int16'
[1632245964135] [87111:7198158] CHIP: [TOO] Error: maxMeasuredValue is lower than expected. Min value is -27314 but got '-32768'

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
What is being fixed?  Examples:
* MaxMeasuredValue needs to be between -27314 to 32767 , but SDK is reporting -32768.

* Fixes #9864

#### Change overview
Fix MaxMeasuredValue is out of range Temperature Measurement Cluster

#### Testing
How was this tested? (at least one bullet point required)
```
yufengw@yufengw-SEi:~/connectedhomeip/out/debug/standalone$ ./chip-tool temperaturemeasurement read measured-value 
.....
[1646333860.006585][34569:34574] CHIP:DMG: ReportDataMessage =
[1646333860.006589][34569:34574] CHIP:DMG: {
[1646333860.006592][34569:34574] CHIP:DMG: 	AttributeReportIBs =
[1646333860.006598][34569:34574] CHIP:DMG: 	[
[1646333860.006601][34569:34574] CHIP:DMG: 		AttributeReportIB =
[1646333860.006606][34569:34574] CHIP:DMG: 		{
[1646333860.006610][34569:34574] CHIP:DMG: 			AttributeDataIB =
[1646333860.006614][34569:34574] CHIP:DMG: 			{
[1646333860.006618][34569:34574] CHIP:DMG: 				DataVersion = 0x9651a08a,
[1646333860.006622][34569:34574] CHIP:DMG: 				AttributePathIB =
[1646333860.006626][34569:34574] CHIP:DMG: 				{
[1646333860.006630][34569:34574] CHIP:DMG: 					Endpoint = 0x1,
[1646333860.006634][34569:34574] CHIP:DMG: 					Cluster = 0x402,
[1646333860.006638][34569:34574] CHIP:DMG: 					Attribute = 0x0000_0000,
[1646333860.006642][34569:34574] CHIP:DMG: 				}
[1646333860.006646][34569:34574] CHIP:DMG: 					
[1646333860.006650][34569:34574] CHIP:DMG: 					Data = NULL
[1646333860.006654][34569:34574] CHIP:DMG: 			},
[1646333860.006658][34569:34574] CHIP:DMG: 			
[1646333860.006662][34569:34574] CHIP:DMG: 		},
[1646333860.006666][34569:34574] CHIP:DMG: 		
[1646333860.006670][34569:34574] CHIP:DMG: 	],
[1646333860.006674][34569:34574] CHIP:DMG: 	
[1646333860.006678][34569:34574] CHIP:DMG: 	SuppressResponse = true, 
[1646333860.006681][34569:34574] CHIP:DMG: 	InteractionModelRevision = 1
[1646333860.006684][34569:34574] CHIP:DMG: }
[1646333860.006739][34569:34574] CHIP:TOO: Endpoint: 1 Cluster: 0x0000_0402 Attribute 0x0000_0000DataVersion: 2521931914
[1646333860.006759][34569:34574] CHIP:TOO:   MeasuredValue: null

```

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: src/app/clusters/temperature-measurement-server/temperature-measurement-server.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+    if (MeasuredValue::Get(endpoint, MaxMeasuredValue) != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        MaxMeasuredValue.SetNull();
+    }
+
+    switch (attributePath.mAttributeId)
+    {
+    case MeasuredValue::Id: {
+        requested = static_cast<int16_t>(chip::Encoding::LittleEndian::Get16(value));
```
Comment (bzbarsky-apple) [inline src/app/clusters/temperature-measurement-server/temperature-measurement-server.cpp:75]: This is wrong, I believe.  `value` is native-endian, not little-endian, as far as I know.  Same for the other attributes.

File: src/app/clusters/temperature-measurement-server/temperature-measurement-server.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        if (requested < (MinMeasuredValue.IsNull() ? kMinMeasuredValueLowerLimit : MinMeasuredValue.Value()))
```
Comment (bzbarsky-apple) [inline src/app/clusters/temperature-measurement-server/temperature-measurement-server.cpp:77]: We should pull `ValueOr` into `Nullable` and use that here.  Followup fine for this part.

File: src/app/clusters/temperature-measurement-server/temperature-measurement-server.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            return Status::InvalidValue;
+        }
+        else if (requested > (MaxMeasuredValue.IsNull() ? kMaxMeasuredValueUpperLimit : MaxMeasuredValue.Value()))
```
Comment (bzbarsky-apple) [inline src/app/clusters/temperature-measurement-server/temperature-measurement-server.cpp:81]: This is going to disallow setting the attribute to null, no?  Was this tested?  Same for the other attributes.

File: src/app/clusters/temperature-measurement-server/temperature-measurement-server.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+constexpr int16_t kMaxMeasuredValueUpperLimit = 32767;  // 327.67°C
+constexpr int16_t kMaxToleranceValue          = 2048;
+
+// =============================================================================
+// Pre-change callbacks for cluster attributes
+// =============================================================================
+
+using Status = Protocols::InteractionModel::Status;
+
+Protocols::InteractionModel::Status MatterTemperatureMeasurementClusterServerPreAttributeChangedCallback(
```
Comment (bzbarsky-apple) [inline src/app/clusters/temperature-measurement-server/temperature-measurement-server.cpp:53]: I am wondering why this is here at all.  The attributes are not writable.  Are we trying to protect the app author from accidentally setting an invalid value?  But this is also forcing them to set min/max/measured in the right order, which is not documented anywhere....

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Comment by yufengwangca** [comment]
Per sync offline, we will not enforce the range check in SDK, but just assume that the device maker will not set the invalid value. All temperature attributes shall be null if their value is unknown.

**Review by bzbarsky-apple — APPROVED** [review]
Thank you!

**Review by andy31415 — APPROVED** [review]

**Pull request merged**